timemachine/fe/free_energy.py

```diff
@@ -340,14 +351,49 @@ def sample(initial_state: InitialState, md_params: MDParams, max_buffer_frames:
         store_x_interval=0,
     )
 
+    rng = np.random.default_rng(md_params.n_frames * md_params.steps_per_frame)
+
     assert np.all(np.isfinite(ctxt.get_x_t())), "Equilibration resulted in a nan"
 
     def run_production_steps(n_steps: int) -> Tuple[NDArray, NDArray]:
-        _, coords, boxes = ctxt.multiple_steps_U(
-            n_steps=n_steps,
-            store_u_interval=0,
-            store_x_interval=md_params.steps_per_frame,
-        )
+        if md_params.local_steps <= 0:
+            _, coords, boxes = ctxt.multiple_steps_U(
+                n_steps=n_steps,
+                store_u_interval=0,
+                store_x_interval=md_params.steps_per_frame,
+            )
+        else:
+            coords = None
+            boxes = None
+            ctxt.setup_local_md(initial_state.integrator.temperature, md_params.freeze_reference)
+            for steps in batches(n_steps, md_params.steps_per_frame):
+                if steps < md_params.steps_per_frame:
+                    warn(
+                        f"Batch of sample has {steps}, less than batch size {md_params.steps_per_frame}. Setting to {md_params.steps_per_frame}"
+                    )
+                    steps = md_params.steps_per_frame
+                global_steps = steps - md_params.local_steps
+                local_steps = md_params.local_steps
+                if global_steps > 0:
+                    ctxt.multiple_steps(
+                        n_steps=global_steps,
+                    )
+                x_t, box_t = ctxt.multiple_steps_local(
+                    local_steps,
+                    initial_state.ligand_idxs.astype(np.int32),
+                    k=md_params.k,
+                    radius=rng.uniform(md_params.min_radius, md_params.max_radius),
+                    seed=rng.integers(np.iinfo(np.int32).max),
+                    burn_in=0,
+                )
+
+                if coords is None:
+                    coords = np.array(x_t)
+                    boxes = np.array(box_t)
+                else:
+                    coords = np.concatenate([coords, x_t])
+                    boxes = np.concatenate([boxes, box_t])
```

Comment:
nit: consider splitting this `else` block into a separate function (`run_production_steps_local`?)

Reply:
Split in https://github.com/proteneer/timemachine/pull/1026/commits/b18d1e0bd6b82a41b9b20f77182656b126502f97